### PR TITLE
itest: use `HarnessTest` to manage the tests

### PR DESCRIPTION
**Pull request by yyforyongyu**
~~This is a follow up PR after and depends on #5756. The first 9 commits are irrelevant for this PR.~~

This PR deals with some long-standing issues, such as context leaks, unclear structs, DRYs, by refactoring the tests into a SOLID-ish style. It also revives the old idea of `HarnessTest`, which provides a nice abstraction to assist us in writing better tests. Our flakes seen so far have complicated origins and it’s likely we would have more as we keep writing tests. Thus we want to “harness” the harness test to increase the maintainability.

The changes made in this PR will make it a happier experience when writing tests, also easier to catch flakes. It enables us to have more control and paves the way for future optimization. See the follow-up  PR #5940.

In short, this PR introduces `HarnessTest` and uses it as an assertion machine. See [README](https://github.com/lightningnetwork/lnd/blob/155fd0db2febea45befb9d56ab2039f951039e9d/lntest/README.md) for more details.

Apologies for making such a large PR, as we have so many test cases, there’s probably no easier way if we want to keep the CI running for all test cases inside this PR.

### CODE REVIEW

File: lntest/harness.go [hunk 1]
```diff
@@ -66,3 +84,478 @@ func (h *HarnessTest) FailNow() {
...
+				Pubkey: bInfo.IdentityPubkey,
+				Host:   b.Cfg.P2PAddr(),
+			},
+		}
+
+		ctxt, cancel := context.WithTimeout(h.runCtx, DefaultTimeout)
+		defer cancel()
+
+		var predErr error
+		err := wait.Predicate(func() bool {
```
Comment (guggero) [inline lntest/harness.go:426]: Use `wait.NoError` instead?

File: lntest/itest/lnd_rest_api_test.go [hunk 2]
```diff
@@ -116,20 +115,15 @@ func testRestAPI(net *lntest.NetworkHarness, ht *harnessTest) {
...
 		run: func(t *testing.T, a, b *lntest.HarnessNode) {
-			// Get a new wallet address from Alice.
-			ctxb := context.Background()
-			newAddrReq := &lnrpc.NewAddressRequest{
-				Type: lnrpc.AddressType_WITNESS_PUBKEY_HASH,
-			}
-			addrRes, err := a.NewAddress(ctxb, newAddrReq)
-			require.Nil(t, err, "get address")
+			// Use a fake address.
+			addr := "bcrt1qlutnwklt4u2548cufrjmsjclewugr9lcpnkzag"
```
Comment (guggero) [inline lntest/itest/lnd_rest_api_test.go:119]: Is there a specific reason not to create a new address from the node here?

File: lntest/itest/lnd_rpc_middleware_interceptor_test.go [hunk 3]
```diff
@@ -492,23 +472,24 @@ type middlewareHarness struct {
...
-func registerMiddleware(t *testing.T, node *lntest.HarnessNode,
+func registerMiddleware(ht *lntest.HarnessTest, node *lntest.HarnessNode,
 	registration *lnrpc.MiddlewareRegistration) *middlewareHarness {
 
-	ctxc, cancel := context.WithCancel(context.Background())
+	// ctxc, cancel := context.WithCancel(context.Background())
 
-	middlewareStream, err := node.RegisterRPCMiddleware(ctxc)
-	require.NoError(t, err)
+	// middlewareStream, err := node.RegisterRPCMiddleware(ctxc)
```
Comment (guggero) [inline lntest/itest/lnd_rpc_middleware_interceptor_test.go:480]: nit: commented out code.

**Review by guggero — COMMENTED** [review]
Absolutely amazing work on this PR! So many old patterns can finally be removed :100:
Thank you so much for tackling this, I can only imagine how much nerves and time this took...

I was able to review about 1/4 of the commits, will definitely need a few more rounds. The PR is massive but I understand that it's not easy to do this in chunks. Also I tried to refrain from nit comments at this point to not block this huge refactor unnecessarily.

File: lntest/harness.go (hunk 1 above)
Comment (yyforyongyu) [reply lntest/harness.go:426]: Changed.

File: lntest/harness.go [hunk 4]
```diff
@@ -66,3 +84,478 @@ func (h *HarnessTest) FailNow() {
...
+	var (
+		mempool []*chainhash.Hash
+		err     error
+	)
+
+	err = wait.NoError(func() error {
+		// We require the RPC call to be succeeded and won't wait for
+		// it as it's an unexpected behavior.
+		mempool, err = h.net.Miner.Client.GetRawMempool()
+		require.NoError(h, err, "unable to get mempool")
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/harness.go]: Make sense, switched!

File: lntest/harness.go [hunk 5]
```diff
@@ -557,6 +557,30 @@ func (h *HarnessTest) SendPayment(hn *HarnessNode,
...
+		num := len(chanInfo.Channels)
+		if num != numChannels {
+			predErr = fmt.Errorf("expected %v channels, got %v",
+				numChannels, num)
+			return false
+		}
+		return true
+	}
+
+	require.NoErrorf(h, wait.Predicate(pred, DefaultTimeout),
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/harness.go:1098]: done.

File: lntest/harness.go (hunk 1 above)
Comment (yyforyongyu) [reply lntest/harness.go:426]: Done

File: lntest/itest/lnd_multi-hop_htlc_receiver_chain_claim_test.go [hunk 6]
```diff
@@ -169,97 +132,69 @@ func testMultiHopReceiverChainClaim(net *lntest.NetworkHarness, t *harnessTest,
...
 	// Carol's pending channel report should now show two outputs under
 	// limbo: her commitment output, as well as the second-layer claim
 	// output.
-	pendingChansRequest := &lnrpc.PendingChannelsRequest{}
-	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
-	pendingChanResp, err := carol.PendingChannels(ctxt, pendingChansRequest)
-	require.NoError(t.t, err)
-
-	require.NotZero(t.t, len(pendingChanResp.PendingForceClosingChannels))
-	forceCloseChan := pendingChanResp.PendingForceClosingChannels[0]
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_multi-hop_htlc_receiver_chain_claim_test.go:200]: Done! Tbh `AssertHTLCStage` is not good enough, something I want to do in the future is to give each of the functions a more explicit name and, make sure we have to ability to assert a single state and not check multiples like the current ones do.

File: lntest/itest/lnd_rpc_middleware_interceptor_test.go (hunk 3 above)
Comment (yyforyongyu) [reply lntest/itest/lnd_rpc_middleware_interceptor_test.go:480]: Fixed!

File: lntest/itest/lnd_rest_api_test.go (hunk 2 above)
Comment (yyforyongyu) [reply lntest/itest/lnd_rest_api_test.go:119]: Yeah to inline with other test cases here since they use fake data.

**Comment by yyforyongyu** [comment]
Thank you so much for reviewing this PR @guggero !! So the new change now makes sure we only use `wait.NoError`. May need to add another commit to remove some old usage of `require.Eventually` and `wait.Predicate`, which will be done in its following PR where more code clean is approached.

File: lntest/itest/lnd_forward_interceptor_test.go [hunk 7]
```diff
@@ -29,166 +31,136 @@ type interceptorTestCase struct {
...
+		payAddr:    invoice.PaymentAddr,
+	}
+
+	// We initiate a payment from Alice.
+	done := make(chan struct{})
+	go func() {
+		ctx.sendPaymentAndAssertAction(tc)
+
+		// Signal that all the payments have been sent.
+		close(done)
```
Comment (guggero) [inline lntest/itest/lnd_forward_interceptor_test.go:67]: This should be a `defer close(done)` since any `require` failure inside `sendPaymentAndAssertAction` will exit the goroutine with only calling `defer` statements.

File: lntest/itest/lnd_forward_interceptor_test.go [hunk 8]
```diff
@@ -29,166 +31,136 @@ type interceptorTestCase struct {
...
+		case ok && status.Code() == codes.Canceled:
+			fallthrough
+
+		// When the test ends, during the node's shutdown it will close
+		// the connection.
+		case strings.Contains(err.Error(), "closed network connection"):
+			fallthrough
+
+		case strings.Contains(err.Error(), "EOF"):
+			close(errDone)
```
Comment (guggero) [inline lntest/itest/lnd_forward_interceptor_test.go:133]: Same here concerning `defer` and use of `require` inside goroutines.

File: lntest/itest/lnd_forward_interceptor_test.go [hunk 9]
```diff
@@ -204,266 +200,161 @@ func testForwardInterceptorDedupHtlc(net *lntest.NetworkHarness, t *harnessTest)
...
-
-			case routerrpc.ResolveHoldForwardAction_RESUME:
-				require.Equal(t.t, lnrpc.HTLCAttempt_SUCCEEDED,
-					attempt.Status, "expected payment to succeed")
-			}
+			ctx.sendPaymentAndAssertAction(tc)
 		}
+
+		// Signal that all the payments have been sent.
+		close(done)
```
Comment (guggero) [inline lntest/itest/lnd_forward_interceptor_test.go:204]: Use `defer` (see other comments).

File: lntest/itest/lnd_forward_interceptor_test.go [hunk 10]
```diff
@@ -204,266 +200,161 @@ func testForwardInterceptorDedupHtlc(net *lntest.NetworkHarness, t *harnessTest)
...
+		case ok && status.Code() == codes.Canceled:
+			fallthrough
+
+		// When the test ends, during the node's shutdown it will close
+		// the connection.
+		case strings.Contains(err.Error(), "closed network connection"):
+			fallthrough
+
+		case strings.Contains(err.Error(), "EOF"):
+			close(done)
```
Comment (guggero) [inline lntest/itest/lnd_forward_interceptor_test.go:304]: Use `defer` in combination with `require` (see other comments).

File: lntest/itest/lnd_network_test.go [hunk 11]
```diff
@@ -118,98 +117,71 @@ func testReconnectAfterIPChange(net *lntest.NetworkHarness, t *harnessTest) {
...
 	// ip1 derived above.
-	dave := net.NewNode(t.t, "Dave", daveArgs, withP2PPort(ip1))
-	defer shutdownAndAssert(net, t, dave)
-
-	// Subscribe to graph notifications from Charlie so that we can tell
-	// when he receives Dave's NodeAnnouncements.
-	ctxb := context.Background()
-	charlieSub := subscribeGraphNotifications(ctxb, t, charlie)
-	defer close(charlieSub.quit)
+	dave := ht.NewNode("Dave", daveArgs)
```
Comment (guggero) [inline lntest/itest/lnd_network_test.go:122]: I wonder why this test ever worked, given we tried to connect to a port that Dave never listened on?

**Review by guggero — COMMENTED** [review]
Second round. I think I got through about half of the changes now...

Needs a rebase too.

File: lntest/itest/lnd_forward_interceptor_test.go (hunk 7 above)
Comment (yyforyongyu) [reply lntest/itest/lnd_forward_interceptor_test.go]: Good catch, now fixed!

File: lntest/itest/lnd_forward_interceptor_test.go (hunk 8 above)
Comment (yyforyongyu) [reply lntest/itest/lnd_forward_interceptor_test.go]: Fixed!

File: lntest/itest/lnd_forward_interceptor_test.go (hunk 9 above)
Comment (yyforyongyu) [reply lntest/itest/lnd_forward_interceptor_test.go]: Fixed!

File: lntest/itest/lnd_forward_interceptor_test.go (hunk 10 above)
Comment (yyforyongyu) [reply lntest/itest/lnd_forward_interceptor_test.go]: Fixed!

File: lntest/itest/lnd_network_test.go (hunk 11 above)
Comment (yyforyongyu) [reply lntest/itest/lnd_network_test.go]: Changed the test a bit to make it more explicit. When we create a new node, we'd auto-assign a p2p port.

**Comment by yyforyongyu** [comment]
> Second round. I think I got through about half of the changes now...
> 
> Needs a rebase too.

Thanks again!

**Comment by guggero** [comment]
Removed my request for review until after 0.15 is shipped. Then will prio this!

**Comment by Roasbeef** [comment]
+1 on some focused attention in this area after 0.15 ships!

**Comment by yyforyongyu** [comment]
Replaced by #6825.
